### PR TITLE
[cudagraph] Warm up two optimizer steps before capture

### PR DESCRIPTION
**Pull request by xmfan**
Stacked PRs:
 * #4485
 * #4435
 * __->__#4433
 * #4432
 * #4430

--- --- ---

## Why

Only the forward-backward callable is captured; the optimizer runs eagerly. Warmup gives the forward-backward path an opportunity to perform lazy initialization before capture and lets the first optimizer step initialize its state. The Trainer zeroes existing gradients in place when CUDA graphs are enabled, preserving their storage for replay.

Two full optimizer steps are a conservative default: the second exercises an eager forward-backward invocation after the first optimizer update. One full optimizer step may be sufficient. We have not established a failure that requires a second step, and a fixed warmup count does not guarantee that every configuration has completed all lazy initialization.

One forward-backward invocation is also not one optimizer step: gradient accumulation runs several invocations per step, and scheduled SDC replay re-executes checked invocations.

## Capture timing

```text
Optimizer step 1
    +--> gradient-accumulation forward/backward calls
    +--> scheduled SDC replay calls
    +--> lazy optimizer-state allocation

Optimizer step 2
    +--> gradient-accumulation forward/backward calls
    +--> scheduled SDC replay calls
    +--> another eager step after the first optimizer update

Optimizer step 3
    +--> capture the next forward/backward invocation
    +--> replay thereafter
```

The warmup count is measured in wrapped forward-backward invocations:

```text
2 * gradient_accumulation_steps
+ checked_warmup_steps * sdc_num_replays
```

## Change

Add a `num_warmup_iterations` option to `CUDAGraphWrapper`, validated in its constructor. The wrapper counts eager forward-backward invocations before capture. The default training policy allows two complete optimizer steps, including gradient accumulation and additional SDC replay calls, before capture starts.

## Testing

The tests cover explicit and invalid warmup counts, gradient accumulation, finite and continuous SDC schedules, and forwarding the configured count into the graph wrapper.

- `python -m pytest -q tests/unit_tests/cpu/test_cudagraph.py tests/unit_tests/cpu/test_trainer.py`
- `python -m tests.integration_tests.run_tests <output> --test_suite features --test_name sdc_replay_cudagraph --execution_mode real_pg --ngpu 1 --no-parallel` (capture begins on step 3)
- Full stack: four-GPU `pp_looped_1f1b_cudagraph` integration, eager steps 1-2, capture on step 3, replay through step 10

### CODE REVIEW

File: torchtitan/trainer.py [hunk 1]
```diff
@@ -676,6 +681,20 @@ def __init__(self, config: Config):
             f"(warmup {config.lr_scheduler.warmup_steps})"
         )
 
+    def _num_cuda_graph_warmup_iterations(self) -> int:
+        num_warmup_steps = 2
+        num_iterations = num_warmup_steps * self.gradient_accumulation_steps
```
Comment (xmfan) [inline torchtitan/trainer.py:686]: This is a bit awkward. So we can't wrap train step (lots of non-cudagraphable stuff), and must wrap the inner fwd bwd function. So when we want to warm up a number of train steps, we have to compute the number of cudagraph wrapped skips

File: torchtitan/trainer.py [hunk 2]
```diff
@@ -676,6 +681,22 @@ def __init__(self, config: Config):
             f"(warmup {config.lr_scheduler.warmup_steps})"
         )
 
+    def _num_cuda_graph_warmup_iterations(self) -> int:
```
Comment (tianyu-l) [inline torchtitan/trainer.py:684]: I'd prefer this happens inside cudagraph.py. E.g. we can send `gradient_accumulation_steps`, `sdc_config.num_steps`, and `sdc_config.num_replays` into `wrap_with_cuda_graph`.

File: torchtitan/trainer.py
Comment (tianyu-l) [inline torchtitan/trainer.py:851]: regarding
> The captured backward writes into param.grad at baked-in addresses. With CUDA graphs enabled the Trainer zeroes gradients in place (set_to_none=False), so gradient buffers are allocated once, by the first eager backward. Capturing earlier would allocate them inside the capture as graph-pool storage that the eager optimizer then reads behind the graph's back.

Due to memory concerns, I think we'll need to apply @BoyuanFeng 's trick to not allocating cudagraph buffer for FSDP-sharded gradients. But yeah we probably still need warmup for other reasons.

File: torchtitan/trainer.py [hunk 3]
```diff
@@ -676,6 +681,22 @@ def __init__(self, config: Config):
...
+        num_warmup_steps = 2
```
Comment (tianyu-l) [inline torchtitan/trainer.py:685]: > One optimizer step is not enough: step 1 performs the one-time allocations above, so the steady-state per-step cycle (gradients accumulated then zeroed in place, optimizer state updated in place) only starts repeating from step 2. Running two full steps eagerly and capturing on step 3 guarantees the recorded invocation matches every subsequent step.

I might have missed something -- why warming up 2 steps instead of just 1, i.e. what makes step 2 and step 3 different?

File: torchtitan/trainer.py
Comment (xmfan) [reply torchtitan/trainer.py:851]: Yeah that will be this row https://docs.google.com/spreadsheets/d/1ELMhvKn3i8xeD-y7D5T3upqS94eu9AJuswyWv9ek53U/edit?gid=0#gid=0

File: torchtitan/trainer.py (hunk 3 above)
Comment (xmfan) [reply torchtitan/trainer.py:685]: 1 step warmup can be sufficient, but it's implementation dependent, and I'm not sure if @sanketpurandare actually hit an issue there.

In theory, you could have lazy state that initializes only the step after optimizer states are initializes, and changes the memory allocation pattern. But if there's none, 1 step should be fine.
